klog_test.go

```diff
@@ -1852,3 +1856,38 @@ func (s *structWithLock) addWithDefer() {
 	defer s.m.Unlock()
 	s.n++
 }
+
+func TestFlushDaemon(t *testing.T) {
+	for sev := severity.InfoLog; sev < severity.FatalLog; sev++ {
+		var flushed int
+		spyFunc := func() { flushed++ }
+		testLog := loggingT{flushD: newFlushDaemon(10*time.Millisecond, spyFunc)}
+
+		// Calling testLog will call createFile, which should start the daemon.
+		testLog.print(sev, nil, nil, "x")
+
+		if !testLog.flushD.isRunning() {
+			t.Error("expected flushD to be running")
+		}
+
+		time.Sleep(400 * time.Millisecond)
```

Comment:
Comment on why sleep?  This is usually a smell...

Reply:
Or inject a fake ticket so you can manually "advance time" rather than sleep.  Deterministic tests FTW

Reply:
Thanks for your feedback. I'm not quite sure what you mean by "fake ticket". Did you mean ticker? I think the problem here is the startup time of the goroutine, not the flush interval, so I don't understand how a fake ticker could help with that.
But I agree that tests should be deterministic if possible and I'd be glad if you could give me some more hints on how to make it deterministic. :slightly_smiling_face:

Reply:
Sorry, I meant "ticker".

e.g. staging/src/k8s.io/client-go/util/workqueue/rate_limiting_queue_test.go - it uses `testingclock.NewFakeClock` (k8s.io/utils/clock/testing) which has a `NewTicker` method.

Pass a `k8s.io/utils/clock.WithTicker` to `newFlushDaemon()` (passing `nil` means "use time.NewTicker()") and store that  in the `flushDaemon` struct.  In `run()` you can check if the passed-in `clock.WithTicker` is nil.  If so, do what this does now (use a real ticker) but if it is not nil, use it instead.

Then your test amnipulates the fake clock's time to cause ticks when you expect them.

As for the daemon startup - you already call isRunning(), so you know it's alive...

Reply:
I've added the clock, thanks for the great tip.
In addition, I had to move the creation of the ticker out of the go function and add a channel for communication between the spy function and the test.